Fixes #837

nilay77 opened the pull request:
- Fixed Graph.subjects, Graph.predicates, and Graph.objects to return only unique values. 
- Added test case for the above 3 methods

niklasl commented on rdflib/graph.py:
```diff
@@ -617,18 +617,27 @@ def set(self, triple):
 
     def subjects(self, predicate=None, object=None):
         """A generator of subjects with the given predicate and object"""
+        subs = []
```
Use sets for keeping track of seen (lists have O(n) lookup time).
```suggestion
        subs = set()
```

(And subsequently use `subs.add(s)`).

nilay77 replied:
Oh yes, this makes sense. I'll update this. Are any other changes required?

nilay77 replied:
@niklasl I have pushed the suggested changes.

tgbugs commented:
This cannot go into core as is because the memory usage will explode for large graphs. As mentioned in the discussion of #837 this is something that probably has to be implemented on the stores or under a different name, or with a different set of arguments e.g. `def subjects(self, predicate=None, object=None, unique=False)` or `def subjects_unique(self, predicate=None, object=None)`.

nilay77 commented:
@tgbugs I have added a parameter named unique to the current methods as suggested by you. I have also added exception handling code which suggests setting the parameter unique to False. If you could give me an idea about how to produce a memory error I can even add a test case.

nicholascar commented on rdflib/graph.py:
```diff
@@ -615,20 +615,56 @@ def set(self, triple):
+    def subjects(self, predicate=None, object=None, unique=False):
+        """A generator of unique subjects with the given predicate and object"""
+
+        if(unique==False):
```
change to `if not unique:`

nicholascar commented on rdflib/graph.py:
```diff
@@ -615,20 +615,56 @@ def set(self, triple):
+    def predicates(self, subject=None, object=None, unique=False):
+        """A generator of unique predicates with the given subject and object"""
+
+        if(unique==False):
```
change to `if not unique:`

nicholascar commented on rdflib/graph.py:
```diff
@@ -615,20 +615,56 @@ def set(self, triple):
+    def objects(self, subject=None, predicate=None, unique=False):
+        """A generator of unique objects with the given subject and predicate"""
+
+        if(unique==False):
```
change to `if not unique:`

nicholascar commented:
Closing this PR as the proponent hasn't addressed comments in a couple of months. @nilay77 please reopen if you want to progress this, but you will need to address comments.
